legacy/src/main/java/com/amazon/opendistroforelasticsearch/sql/legacy/plugin/RestSQLQueryAction.java

```diff
@@ -109,10 +124,15 @@ public RestChannelConsumer prepareRequest(SQLQueryRequest request, NodeClient no
       return NOT_SUPPORTED_YET;
     }
 
-    if (request.isExplainRequest()) {
-      return channel -> sqlService.explain(plan, createExplainResponseListener(channel));
+    try {
+      if (request.isExplainRequest()) {
+        return channel -> sqlService.explain(plan, createExplainResponseListener(channel));
+      }
+      return channel -> sqlService.execute(plan, createQueryResponseListener(channel));
+    } catch (Exception e) {
```

Comment:
As I understand, `SQLService` will execute/explain the request by `ElasticsearchExecutionEngine` which delegates to ES worker thread. In this case, this `try-catch` won't capture the exception (if I recall right ES's listener will intercepts and writes response so control flow won't come back here).

In my opinion, we should create a metric collector that implements `ResponseListener`. The metric collector and response formatter forms a chain which is passed to execution engine and handles response/error in Chain of Responsibility style.